src/DiagnosticsManager.ts

```diff
@@ -92,108 +100,161 @@ export class DiagnosticsManager implements vscode.Disposable {
-            );
+            const currentDiagnostic = mergedDiagnostics.find(d => isEqual(d, diagnostic));
             if (currentDiagnostic) {
-                combinedDiagnostics.splice(combinedDiagnostics.indexOf(currentDiagnostic), 1);
+                mergedDiagnostics.splice(mergedDiagnostics.indexOf(currentDiagnostic), 1);
             }
 
             // Perform de-duplication
             if (precedence.includes(diagnostic.source || "")) {
-                combinedDiagnostics.push(diagnostic);
+                mergedDiagnostics.push(diagnostic);
                 continue;
             }
             if (!currentDiagnostic || !precedence.includes(currentDiagnostic.source || "")) {
-                combinedDiagnostics.push(diagnostic);
+                mergedDiagnostics.push(diagnostic);
                 continue;
             }
-            combinedDiagnostics.push(currentDiagnostic);
+            mergedDiagnostics.push(currentDiagnostic);
         }
     }
 
     private removeSwiftcDiagnostics() {
         this.diagnosticCollection.forEach((uri, diagnostics) => {
             const newDiagnostics = diagnostics.slice();
-            this.removeDiagnostics(newDiagnostics, DiagnosticsManager.swiftc);
+            this.removeDiagnostics(newDiagnostics, d => this.isSwiftc(d));
             if (diagnostics.length !== newDiagnostics.length) {
                 this.diagnosticCollection.set(uri, newDiagnostics);
             }
         });
     }
 
-    private removeSourceKitDiagnostics() {
-        this.diagnosticCollection.forEach((uri, diagnostics) => {
-            const newDiagnostics = diagnostics.slice();
-            this.removeDiagnostics(newDiagnostics, DiagnosticsManager.sourcekit);
-            if (diagnostics.length !== newDiagnostics.length) {
-                this.diagnosticCollection.set(uri, newDiagnostics);
+    private isSource(diagnostic: vscode.Diagnostic, sources: string[]): boolean {
+        return sources.includes(diagnostic.source || "");
+    }
+
+    private isSwiftc(diagnostic: vscode.Diagnostic): boolean {
+        return this.isSource(diagnostic, DiagnosticsManager.swiftc);
+    }
+
+    private isSourceKit(diagnostic: vscode.Diagnostic): boolean {
+        return this.isSource(diagnostic, DiagnosticsManager.sourcekit);
+    }
+
+    private removeDiagnostics(
+        diagnostics: vscode.Diagnostic[],
+        matches: (d: vscode.Diagnostic) => boolean
+    ): vscode.Diagnostic[] {
+        const removed: vscode.Diagnostic[] = [];
+        let i = diagnostics.length;
+        while (i--) {
+            if (matches(diagnostics[i])) {
+                removed.push(...diagnostics.splice(i, 1));
             }
-        });
+        }
+        return removed;
     }
 
-    private removeDiagnostics(diagnostics: vscode.Diagnostic[], sources: string[]): void {
+    private getDiagnostics(
+        diagnostics: vscode.Diagnostic[],
+        sources: string[]
+    ): vscode.Diagnostic[] {
```

Comment:
This function modifies its input as well as returns output. Would be more clear if it returned an object of `{ removed: Diagnostic[], diagnostics: Diagnostic[] }`

Reply:
Ya really not needed, I should have just .filter(...) the array